### ISSUE TITLE
axios-request-adapter reaching too far into internals

### ISSUE THREAD

**Issue by carlosdp** [hwchase17/langchainjs#156]
It looks like `axios-request-adapter.js` is importing parts from `axios` that aren't actually exporting from that library in the latest version (`0.0.13`), causing a crash when calling out to APIs using axios:

```
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './lib/core/settle' is not defined by "exports" in /Users/carlosdp/code/weaver/node_modules/axios/package.json imported from /Users/carlosdp/code/weaver/node_modules/langchain/dist/util/axios-fetch-adapter.js
    at new NodeError (node:internal/errors:371:5)
    at throwExportsNotFound (node:internal/modules/esm/resolve:453:9)
    at packageExportsResolve (node:internal/modules/esm/resolve:731:3)
    at packageResolve (node:internal/modules/esm/resolve:913:14)
    at moduleResolve (node:internal/modules/esm/resolve:978:18)
    at defaultResolve (node:internal/modules/esm/resolve:1080:11)
    at ESMLoader.resolve (node:internal/modules/esm/loader:530:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:251:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:79:40)
    at link (node:internal/modules/esm/module_job:78:36) {
  code: 'ERR_PACKAGE_PATH_NOT_EXPORTED'
```

I'll see if I have time to submit a PR for a fix for this later tonight, but downgrading to 0.0.12 is a temporary fix.

That said, I think the **real fix** is to eliminate the `openai` library and just call the HTTP API directly with axios -- it's not immediately clear to me why an adapter should be necessary for what amounts to extremely simple HTTP requests. (Also, the whole "you need to install the openai library" exception after install is a massive anti-pattern, so would be good to eliminate that anyways!)

**Comment by nfcampos** [comment]
Hi @carlosdp thanks for opening the issue. It sounds like the issue might be in your project you have another dependency maybe using`axios` 1.x? The most recent 0.x axis version I can find on npm shouldn't have this issue https://www.npmjs.com/package/axios/v/0.27.2?activeTab=explore

I cannot guarantee our dependencies that use `axios` (like openai and chromadb) are compatible with axios 1.x. In any case I've opened a PR copying the internals over here #163 

We have dynamic imports for our peer dependencies (as specified in our package.json) in order to support the users who do not need all of our dependencies. If you have another suggestion for handling those, let us know, always happy to improve

**Comment by cogscides** [comment]
I have the same issue

**Comment by carlosdp** [comment]
> Hi @carlosdp thanks for opening the issue. It sounds like the issue might be in your project you have another dependency maybe using`axios` 1.x? The most recent 0.x axis version I can find on npm shouldn't have this issue https://www.npmjs.com/package/axios/v/0.27.2?activeTab=explore
> 
> I cannot guarantee our dependencies that use `axios` (like openai and chromadb, which both specify 0.x as the version of axios they want) are compatible with axios 1.x. In any case I've opened a PR copying the internals over here #163
> 
> We have dynamic imports for our peer dependencies (as specified in our package.json) in order to support the users who do not need all of our dependencies. If you have another suggestion for handling those, let us know, always happy to improve

Gotcha! This would be handled by the node package system if they weren't dynamic imports! I would strongly advise moving away from this dynamic-import-runtime-exception-notice pattern for external package dependencies. It's going to result in a lot of GH issues like these moving forward 😅 

I have one recommendation that I promise you will have insane ROI if you do it now rather than wait till the project grows a bunch:

Use the multi-package pattern for non-core classes, instead of the mono-package. Every single JS project that is similar to this (where they have a bunch of adapters) starts out like this, and always ends up having to do a big refactor to break into multi-package. Then, things would look like this:

```typescript
import { LLMChain } from '@langchain/core';
import { OpenAI } from '@langchain/openai';
import { PineconeDB } from '@langchain/pinecone';
...
```

This avoids dynamic imports, but more importantly the developer has a clear pattern to follow when importing their desired integration, and they discover that at `yarn install` time when writing the code, *not* when they try to run the project and it runtime errors. ✨ Much better DevEx! This also avoids errors like these, because each individual integration can have it's own (statically declared) version of axios (if the package system determines that is needed), for example.

**Note:** You can still use a monorepo! Just publish the distinct modules as new npm packages.

Example repo: [Chakra](https://github.com/chakra-ui/chakra-ui)

**Comment by nfcampos** [comment]
Thank for the feedback @carlosdp, we'll take it into consideration. Closing this issue as the original issue has been resolved in #163
